math/binary_exponent.cpp

```diff
@@ -50,22 +97,38 @@ int binExpo_alt(int a, int b) {
     }
     return res;
 }
+}  // namespace math
 
-/// Main function
-int main() {
-    int a, b;
-    /// Give two numbers a, b
-    std::cin >> a >> b;
-    if (a == 0 && b == 0) {
-        std::cout << "Math error" << std::endl;
-    } else if (b < 0) {
-        std::cout << "Exponent must be positive !!" << std::endl;
-    } else {
-        int resRecurse = binExpo(a, b);
-        /// int resIterate = binExpo_alt(a, b);
+/**
+ * @brief Self-test implementation
+ * @returns void
+ */
+static void tests() {
+    assert(math::binExpo_recursive(1, 0) == 1);
+    assert(math::binExpo_recursive(746584, 0) == 1);
+    assert(math::binExpo_recursive(1, 1) == 1);
+    assert(math::binExpo_recursive(2938374, 1) == 2938374);
+    assert(math::binExpo_recursive(3, 7) == 2187);
+    assert(math::binExpo_recursive(31, 5) == 28629151);
+    assert(math::binExpo_recursive(0, 0) == 1);
+    assert(math::binExpo_recursive(1, -20) == 0);
 
-        /// Result of a^b (where '^' denotes exponentiation)
-        std::cout << resRecurse << std::endl;
-        /// std::cout << resIterate << std::endl;
-    }
+    assert(math::binExpo_iterative(1, 0) == 1);
+    assert(math::binExpo_iterative(746584, 0) == 1);
+    assert(math::binExpo_iterative(1, 1) == 1);
+    assert(math::binExpo_iterative(2938374, 1) == 2938374);
+    assert(math::binExpo_iterative(3, 7) == 2187);
+    assert(math::binExpo_iterative(31, 5) == 28629151);
+    assert(math::binExpo_iterative(0, 0) == 1);
+    assert(math::binExpo_iterative(1, -20) == 0);
+    std::cout << "All tests have successfully passed!" << std::endl;
```

Comment:
Can you consider adding a benchmark to showcase the differences between these two approaches when it comes to speed and memory consumption? Not necessary for me to approve this PR but its a good thing for people learning to use the language.

Reply:
Tried a couple of ways to get a memory usage check for each function, but no success so far.

Considering using getRUsage, but would require download of Cygwin for a ```sys/resource.h```, not sure if that's worthwhile due to users having to download it just to see a simple memory check. 

For now, feel free to approve the PR and maybe make an issue seeing if someone else can do a better job of this then me. If its still up as an issue after a while, I might try again!

Reply:
The proper way to run benchmarks is to run the same function with same inputs over a large number of times rather than doing one call and comparing it to the other call, most computers would run the functions with negligible difference and you will get nothing but if you loop it over a large enough N, then you will have meaningful results.

If you don't think it's possible then feel free to mark this as resolved and then I can look at adding some prior art to doing benchmarks to help people learn😄

Reply:
Sorry I didn't see this until now, I will try to implement it that way at some point and see what the results are. Feel free to do it before me! Can't guarantee a deadline for it.